airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py

```diff
@@ -110,6 +110,14 @@ def read(
                         f"The requested stream {configured_stream.stream.name} was not found in the source."
                         f" Available streams: {stream_instances.keys()}"
                     )
+
+                stream_is_available, error = stream_instance.check_availability(logger, self)
+                if not stream_is_available:
+                    logger.warning(f"Skipped syncing stream '{stream_instance.name}' because it was unavailable. Error: {error}")
+                    continue
+
+                stream_instance.catalog_schema = configured_stream.stream.json_schema
```

Comment:
This feels like an initialization problem and it feels like it's not the first time we struggle with this. The [proposal for data feed](https://docs.google.com/document/d/1UpYo_kPO04KGzDOB7KxBTXUTEzmBB_sBO7X9KddXRBk/edit#heading=h.nav4gq872r1j) was expecting to have the state when we initialize the source but it did not. The consequence was that we had to propagate state everywhere on the `DeclarativeStream.state` setter but as we didn't know all the underlying components, it ended up creating a lot of `if self.cursor and isinstance(self.cursor, <whatever>)`. I fear that because we don't instantiate our component fully at the start of a run, we end-up in this situation where:
* we have multiple place where we manage the initialization
* the internal state of our classes are very hard to understand because the lifecycle isn't just "init everything" but is "init x + init y + ..."

I'm wondering if we can start pushing these up as much as possible close to the initialization of the source. I feel like there are a couple of challenges to do this:
* Not all CDK command provide the same information and therefore fields like `catalog` are not always available
* The source is created before being passed to the entrypoint which finds out some parameters that we would like for the initialization (like `config_catalog = self.source.read_catalog(parsed_args.catalog)`)

I'll think about it. I think the least we could do is put all those initialization at the same place as much as possible. The one I had to deal with was the `state` which is initialize [here](https://github.com/airbytehq/airbyte/blob/e8ddda970905ae338377d2b2a7ed5fa9ffa57ed4/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L239-L242). I'm not sure this is the best place as it is scope to the stream but at least it gives us another case to consider